generic/shared/src/main/scala/io/circe/generic/JsonCodecMacros.scala

```diff
@@ -0,0 +1,50 @@
+package io.circe.generic
+
+import io.circe.{ Decoder, Encoder }
+import macrocompat.bundle
+import scala.language.experimental.macros
+import scala.reflect.macros.blackbox
+
+class JsonCodec extends scala.annotation.StaticAnnotation {
+  def macroTransform(annottees: Any*): Any = macro JsonCodecMacros.jsonCodecAnnotationMacro
+}
+
+@bundle
+private[generic] class JsonCodecMacros(val c: blackbox.Context) {
+  import c.universe._
+
+  def jsonCodecAnnotationMacro(annottees: Tree*): Tree = annottees match {
+    case List(clsDef: ClassDef) if clsDef.mods hasFlag Flag.CASE =>
```

Comment:
This is nitpicky, but I'd like to stick to a no-infix-syntax-for-non-operator-methods rule. That really should be enforced by Scalastyle, though, so we could merge this and fix once a rule is in place and complains about it.